File: ascii_fields/animations/drops.py
import math
import random

from ..core import Animation, clamp, render_field


class _Drop:
  __slots__ = ("x", "y", "t0", "strength")

  def __init__(self, x, y, t0, strength):
    self.x = x
    self.y = y
    self.t0 = t0
    self.strength = strength


class DropsAnimation(Animation):
  """Water drops falling into a pond. Each impact sends out an expanding ring;
  overlapping rings interfere the way real ripples do."""

  thresholds = [
    (0.30, "."), (0.40, ":"), (0.47, "-"), (0.53, "="), (0.60, "+"),
    (0.70, "*"), (0.82, "#"), (1.01, "%"),
  ]

  WAVE_SPEED = 0.55       # rings per second outward (normalised units)
  SPAWN_DT = 0.55

  def __init__(self):
    self._drops = []
    self._next_spawn = 0.0
    self._last_elapsed = 0.0
    self._rng = random.Random()

  def _reset(self):
    self._drops = []
    self._next_spawn = 0.0

  def render(self, width, height, elapsed, phase, options):
    if elapsed < self._last_elapsed:
      self._reset()
    self._last_elapsed = elapsed

    aspect = width / max(1, height * 2.0)
    while elapsed >= self._next_spawn:
      self._drops.append(_Drop(
        self._rng.uniform(0.08, 0.92),
        self._rng.uniform(0.08, 0.92),
        self._next_spawn,
        self._rng.uniform(0.7, 1.0),
      ))
      self._next_spawn += self.SPAWN_DT * self._rng.uniform(0.6, 1.5)
    self._drops = [d for d in self._drops if elapsed - d.t0 < 7.0][-16:]

    contrast = options.contrast
    grid = []
    for row in range(height):
      v = row / max(1, height - 1)
      line = []
      for col in range(width):
        u = col / max(1, width - 1)
        h = 0.0
        for d in self._drops:
          age = elapsed - d.t0
          dx = (u - d.x) * aspect
          dy = v - d.y
          r = math.hypot(dx, dy)
          front = self.WAVE_SPEED * age
          ring = r - front
          envelope = math.exp(-(ring * ring) / 0.010)
          decay = math.exp(-age * 0.5) / (1.0 + 6.0 * r)
          h += d.strength * envelope * decay * math.sin(38.0 * ring)
        line.append(clamp(0.42 + h * 2.4 * contrast))
      grid.append(line)
    return render_field(width, height, grid, options, self)

File: src/animations/drops.rs
use rand::{Rng, SeedableRng};
use rand_pcg::Pcg32;
use crate::animation::{Animation, FrameContext};
use crate::core::{clamp, render_field, FieldStyle};

const STYLE: FieldStyle = FieldStyle { gray_lo: 234, gray_hi: 255, default_theme: "ocean" };
const TH: &[(f64, char)] = &[
  (0.30, '.'), (0.40, ':'), (0.47, '-'), (0.53, '='), (0.60, '+'),
  (0.70, '*'), (0.82, '#'), (1.01, '%'),
];
const WAVE_SPEED: f64 = 0.55;
const SPAWN_DT: f64 = 0.55;

#[derive(Clone, Copy)] struct Drop { x: f64, y: f64, t0: f64, strength: f64 }

pub struct Drops {
  drops: Vec<Drop>,
  next_spawn: f64,
  last_elapsed: f64,
  rng: Pcg32,
}
impl Default for Drops {
  fn default() -> Self {
    Self { drops: Vec::new(), next_spawn: 0.0, last_elapsed: 0.0, rng: Pcg32::from_entropy() }
  }
}

impl Animation for Drops {
  fn render(&mut self, ctx: &FrameContext, out: &mut String) {
    if ctx.elapsed < self.last_elapsed { self.drops.clear(); self.next_spawn = 0.0; }
    self.last_elapsed = ctx.elapsed;
    let ax = ctx.width as f64 / (ctx.height as f64 * 2.0).max(1.0);
    while ctx.elapsed >= self.next_spawn {
      self.drops.push(Drop {
        x: self.rng.gen_range(0.08..0.92),
        y: self.rng.gen_range(0.08..0.92),
        t0: self.next_spawn,
        strength: self.rng.gen_range(0.7..1.0),
      });
      self.next_spawn += SPAWN_DT * self.rng.gen_range(0.6..1.5);
    }
    self.drops.retain(|d| ctx.elapsed - d.t0 < 7.0);
    while self.drops.len() > 16 { self.drops.remove(0); }
    let contrast = ctx.options.contrast;
    let mut grid = vec![0.0_f64; ctx.width * ctx.height];
    let dw = (ctx.width.saturating_sub(1)).max(1) as f64;
    let dh = (ctx.height.saturating_sub(1)).max(1) as f64;
    for row in 0..ctx.height {
      let v = row as f64 / dh;
      let base = row * ctx.width;
      for col in 0..ctx.width {
        let u = col as f64 / dw;
        let mut h = 0.0_f64;
        for d in &self.drops {
          let age = ctx.elapsed - d.t0;
          let dx = (u - d.x) * ax;
          let dy = v - d.y;
          let r = (dx * dx + dy * dy).sqrt();
          let front = WAVE_SPEED * age;
          let ring = r - front;
          let envelope = (-(ring * ring) / 0.010).exp();
          let decay = (-age * 0.5).exp() / (1.0 + 6.0 * r);
          h += d.strength * envelope * decay * (38.0 * ring).sin();
        }
        grid[base + col] = clamp(0.42 + h * 2.4 * contrast);
      }
    }
    render_field(ctx, &grid, TH, &STYLE, out);
  }
}
